include/execution.hpp

```diff
@@ -2252,142 +2180,28 @@ namespace std::execution {
-
-          template <template <class...> class _Variant>
-            using error_types =
-              __result_senders_t<
-                __transform<
-                  __bind_back<__defer<__error_types_of_t>, _Env, __q<__types>>,
-                  __bind_front<
-                    __concat<__munique<__q<_Variant>>>,
-                    __types<exception_ptr>>>>;
-
-          static constexpr bool sends_stopped =
-            __result_senders_t<
-              __transform<
-                __bind_back_q1<__sends_stopped, _Env>,
-                __right_fold<
-                  __sends_stopped<_Sender, _Env>,
-                  __q2<__or>>>>::value;
+      template <class _Env, class _Fun, class _Set, class _Ret, class... _Args>
+          requires (!same_as<_Set, _Ret>)
+        struct __tfx_signal<_Env, _Fun, _Set, _Ret(_Args...)> {
+          using type = completion_signatures<_Ret(_Args...)>;
         };
 
-      template <class _Sender, class _Env, class _Fun>
-          requires sender<_Sender, _Env> &&
-            __valid<__stopped_senders_of, _Sender, _Env, _Fun, __typed_senders<_Env>>
-        struct __completion_sigs<_Sender, _Env, _Fun, set_stopped_t>
-        {
-          template <class _Continuation>
-            using __result_senders_t =
-              __stopped_senders_of<_Sender, _Env, _Fun, _Continuation>;
-
-          template <template <class...> class _Tuple, template <class...> class _Variant>
-            using value_types =
-              __result_senders_t<
-                __transform<
-                  __bind_back<__defer<__value_types_of_t>, _Env, __q<_Tuple>, __q<__types>>,
-                  __bind_front<
-                    __concat<__munique<__q<_Variant>>>,
-                    value_types_of_t<_Sender, _Env, _Tuple, __types>>>>;
-
-          template <template <class...> class _Variant>
-            using error_types =
-              __result_senders_t<
-                __transform<
-                  __bind_back<__defer<__error_types_of_t>, _Env, __q<__types>>,
-                  __bind_front<
-                    __concat<__munique<__q<_Variant>>>,
-                    __types<exception_ptr>,
-                    error_types_of_t<_Sender, _Env, __types>>>>;
-
-          static constexpr bool sends_stopped =
-            __result_senders_t<
-              __transform<
-                __bind_back_q1<__sends_stopped, _Env>,
-                __right_fold<false_type, __q2<__or>>>>::value;
+      template <class _Env, class _Fun, class _Set, class... _Args>
+          requires invocable<_Fun, _Args...> &&
+            sender<invoke_result_t<_Fun, _Args...>, _Env> &&
+            __is_typed_completions<
+              completion_signatures_of_t<invoke_result_t<_Fun, _Args...>, _Env>>
+        struct __tfx_signal<_Env, _Fun, _Set, _Set(_Args...)> {
+          using __completions =
+            completion_signatures_of_t<invoke_result_t<_Fun, _Args...>, _Env>;
+          using type =
```

Comment:
(minor) If I'm reading this correctly we are not properly computing the end completion signatures.

If the input sender completes with `set_value_t(int)`, `set_value_t(string)`, `set_error_t(error_code)`, and the transformation function is  `[](int x) { return just(x + 0.1415); }`, I would expect the output completions to be `set_value_t(int)`, `set_value_t(string)`, `set_error_t(error_code)` (two remain, one removed, one added).

On the other hand, this doesn't seem important at this point

Reply:
No, assuming you're talking about a call to `let_value`, this should not require. The function should be able to handle whatever (sets of) value types with which the input sender can complete. Since `string` doesn't convert to `int`, I wouldn't expect the `connect` call to type-check.